src/hotspot/cpu/x86/vm_version_x86.cpp

```diff
@@ -387,6 +412,52 @@ class VM_Version_StubGenerator: public StubCodeGenerator {
     __ movl(Address(rsi, 8), rcx);
     __ movl(Address(rsi,12), rdx);
 
+#ifndef PRODUCT
+    //
+    // Check if OS has enabled XGETBV instruction to access XCR0
+    // (OSXSAVE feature flag) and CPU supports APX
+    //
+    // To enable APX, check CPUID.EAX=7.ECX=1.EDX[21] bit for HW support
+    // and XCRO[19] bit for OS support to save/restore extended GPR state.
+    __ lea(rsi, Address(rbp, in_bytes(VM_Version::sefsl1_cpuid7_offset())));
+    __ movl(rax, 0x200000);
+    __ andl(rax, Address(rsi, 4));
+    __ cmpl(rax, 0x200000);
+    __ jcc(Assembler::notEqual, vector_save_restore);
+    // check _cpuid_info.xem_xcr0_eax.bits.apx_f
+    __ movl(rax, 0x80000);
+    __ andl(rax, Address(rbp, in_bytes(VM_Version::xem_xcr0_offset()))); // xcr0 bits apx_f
+    __ cmpl(rax, 0x80000);
+    __ jcc(Assembler::notEqual, vector_save_restore);
+
+    /* FIXME: Uncomment after integration of JDK-8328998
+    __ mov64(r15, VM_Version::egpr_test_value());
+    __ mov64(r16, VM_Version::egpr_test_value());
+    __ mov64(r31, VM_Version::egpr_test_value());
+    */
+    __ xorl(rsi, rsi);
+    VM_Version::set_cpuinfo_segv_addr_apx(__ pc());
+    // Generate SEGV
+    __ movl(rax, Address(rsi, 0));
+
+    VM_Version::set_cpuinfo_cont_addr_apx(__ pc());
+    // Validate the contents of r15, r16 and r31
+    /* FIXME: Uncomment after integration of JDK-8328998
+    __ mov64(rax, VM_Version::egpr_test_value());
+    __ cmpq(rax, r15);
+    __ jccb(Assembler::notEqual, apx_save_restore_error);
+    __ cmpq(rax, r16);
+    __ jccb(Assembler::notEqual, apx_save_restore_error);
+    __ cmpq(rax, r31);
+    __ jccb(Assembler::equal, vector_save_restore);
+
+    // Generate SEGV to signal unsuccessful save/restore.
+    __ bind(apx_save_restore_error);
+    __ lea(rax, ExternalAddress(VM_Version::_apx_state_restore_error_handler));
```

Comment:
Generating an error message here won't be the right thing (especially since this is default by feature detection). It should only result in setting UseAPX feature to false.

Reply:
DONE